### ISSUE TITLE
Add AI.EXTRACT with parallel field decoding and shared document KV

### ISSUE THREAD

**Issue by shreyashankar** [fsdatalab/quail#106]
## Problem

`AI.EXTRACT` should extract several independent fields from one document.

Generating one structured response would decode the fields one after another. It would also generate field names and formatting that Quail discards after parsing the result.

Instead, Quail should decode each field independently while sharing the document KV. Field outputs can run in parallel because one field does not depend on another field's output.

Speculative source copying is out of scope and will have a separate issue.

## SQL interface

The user provides a document and a definition for each field:

```sql
SELECT
    r.id,
    AI.EXTRACT(
        r.body,
        {
            'patient_name': {
                'type': 'STRING',
                'question': 'What is the patient name?',
                'max_output_tokens': 32
            },
            'age': {
                'type': 'INT64',
                'question': 'What is the patient age?',
                'max_output_tokens': 8
            },
            'reactions': {
                'type': 'ARRAY<STRING>',
                'question': 'Which reactions did the patient experience?',
                'max_output_tokens': 128
            }
        }
    ) AS patient
FROM reports AS r;
```

The result is an Arrow struct with fields named `patient_name`, `age`, and `reactions`.

Each field has its own question, type, and output limit. The first version supports `STRING`, `INT64`, `FLOAT64`, `BOOL`, and `ARRAY<STRING>`.

A missing scalar value becomes `NULL`. A list with no matching values becomes an empty list.

## Query planning

The logical plan should retain the input document and each field definition. The output field definitions determine the return type before execution.

The planner should delay `AI.EXTRACT` until after filters, joins, and limits that do not use its output. If a later operation uses an extracted field, the planner should run the extraction before that operation and retain the result for later consumers.

The physical plan should compute the document KV once. It should create one decode branch for each field and keep the document KV available until every field finishes.

The cost model should count:

- One document prefill.
- One question suffix for each field.
- The expected generated tokens for each field.
- Attention between each field branch and the shared document KV.
- Attention within each field's question and generated output.

The batch size should include each field's `max_output_tokens`. `query.explain()` should show the field count, field types, output limits, estimated generated tokens, and estimated time.

## Execution

The backend should process all active field branches for one document together.

Each branch contains its field question and generated output. At each decode step, every unfinished branch can generate its next token. A long field should not prevent shorter fields from finishing.

The attention path should separate attention over the field branch from attention over the shared document KV. It should group the field queries so the GPU reads the document KV once for the group instead of once per field.

The backend can extend the attention path used for joins, where several partner suffixes attend to one shared anchor.

The decoder should restrict each field to its declared type. It should return field values directly to the result builder instead of generating a JSON object.

The result builder should combine the completed fields into one Arrow struct.

## Verification

Unit tests should compare the grouped attention output with independent full sequence attention. The logits should match within the tolerance already used for Quail's attention tests.

A Modal benchmark should compare two configurations with the same model, documents, field questions, output limits, and batch settings:

1. Each field runs as an independent request.
2. The field branches run together with shared document KV.

State the prediction before the run. The prediction is that grouped field decoding reduces document KV reads and query time, with a larger benefit as the number of fields or document length increases.

Report query time, documents per second, generated tokens, fresh input tokens, and GPU cost per query. Also report GPU time spent in attention and the bytes of document KV read if the profiler exposes that measurement.

## Acceptance criteria

- `AI.EXTRACT` returns an Arrow struct with the declared fields and types.
- Each field has an independent question and output limit.
- The document prefix is computed once per input document.
- Field branches decode in parallel and share the document KV.
- The grouped attention path matches independent attention within the test tolerance.
- Finished field branches stop consuming decode work.
- The planner delays extraction until its output is needed.
- The planner retains one extraction result for all later consumers.
- The cost model includes every field question and output.
- `query.explain()` shows the extraction and decode estimates.
- Tests cover supported types, missing values, output limits, attention, and result assembly.
- The Modal comparison reports query time, throughput, and GPU cost per query.
